File: backend/features/project_document/formula/tokens.py
"""Token kinds + Token dataclass for the formula tokenizer.

Mirrored by `frontend/.../lib/formula/tokens.ts`. The string enum
values are stable and appear in the shared corpus's `expected_error`
entries.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum


class TokenKind(StrEnum):
    NUMBER = "NUMBER"
    STRING = "STRING"
    BOOL = "BOOL"
    NULL = "NULL"
    IDENT = "IDENT"
    FIELD_REF = "FIELD_REF"
    LPAREN = "LPAREN"
    RPAREN = "RPAREN"
    COMMA = "COMMA"
    PLUS = "PLUS"
    MINUS = "MINUS"
    STAR = "STAR"
    SLASH = "SLASH"
    PERCENT = "PERCENT"
    EQ = "EQ"
    NEQ = "NEQ"
    LT = "LT"
    LTE = "LTE"
    GT = "GT"
    GTE = "GTE"
    AND = "AND"
    OR = "OR"
    NOT = "NOT"
    IF = "IF"
    EOF = "EOF"


@dataclass(frozen=True, slots=True)
class Token:
    kind: TokenKind
    text: str
    offset: int
    # Populated for NUMBER / STRING / BOOL / NULL / FIELD_REF.
    value: str | float | bool | None = None

File: frontend/src/shared/ui/data-table/lib/formula/tokens.ts
// Token kinds + Token interface for the formula tokenizer.
// Mirror of `backend/features/project_document/formula/tokens.py`. The
// `TokenKind` values are stable wire constants — used in error messages
// and the shared corpus.

export const TokenKind = {
  NUMBER: "NUMBER",
  STRING: "STRING",
  BOOL: "BOOL",
  NULL: "NULL",
  IDENT: "IDENT",
  FIELD_REF: "FIELD_REF",
  LPAREN: "LPAREN",
  RPAREN: "RPAREN",
  COMMA: "COMMA",
  AMPERSAND: "AMPERSAND",
  PLUS: "PLUS",
  MINUS: "MINUS",
  STAR: "STAR",
  SLASH: "SLASH",
  PERCENT: "PERCENT",
  EQ: "EQ",
  NEQ: "NEQ",
  LT: "LT",
  LTE: "LTE",
  GT: "GT",
  GTE: "GTE",
  AND: "AND",
  OR: "OR",
  NOT: "NOT",
  IF: "IF",
  EOF: "EOF",
} as const;

export type TokenKind = (typeof TokenKind)[keyof typeof TokenKind];

export type TokenValue = string | number | boolean | null;

export interface Token {
  kind: TokenKind;
  text: string;
  offset: number;
  // Populated for NUMBER / STRING / BOOL / NULL / FIELD_REF.
  value?: TokenValue;
}
